eureka2-write-server/src/test/java/com/netflix/eureka2/server/service/replication/ReplicationHandlerTest.java

```diff
@@ -0,0 +1,414 @@
+        when(messageConnection.incoming()).thenReturn(Observable.<Object>just(HELLO_REPLY));
+
+        TransportClient transportClient = mock(TransportClient.class);
+        when(transportClient.connect()).thenReturn(Observable.just(messageConnection));
+
+        ReplicationChannel channel = new SenderReplicationChannel(transportClient, mock(ReplicationChannelMetrics.class));
+
+        return new TestSenderReplicationChannel(channel, id);
+    }
+
+    private static TestSenderReplicationChannel newFailAtHelloChannel(Integer id) {
+        MessageConnection messageConnection = newMockMessageConnection();
+        when(messageConnection.submit(anyObject())).thenReturn(Observable.<Void>empty());
+        when(messageConnection.incoming()).thenReturn(Observable.error(new Exception("test: hello reply error")));
+
+        TransportClient transportClient = mock(TransportClient.class);
+        when(transportClient.connect()).thenReturn(Observable.just(messageConnection));
+
+        ReplicationChannel channel = new SenderReplicationChannel(transportClient, mock(ReplicationChannelMetrics.class));
+
+        return new TestSenderReplicationChannel(channel, id);
+    }
+
+    private static TestSenderReplicationChannel newFailAtReplicateChannel(Integer id) {
+        MessageConnection messageConnection = newMockMessageConnection();
+        when(messageConnection.submit(any(ReplicationHello.class))).thenReturn(Observable.<Void>empty());
+        when(messageConnection.submit(any(RegisterCopy.class))).thenReturn(Observable.<Void>error(new Exception("test: register error")));
+        when(messageConnection.submit(any(UnregisterCopy.class))).thenReturn(Observable.<Void>error(new Exception("test: unregister error")));
+        when(messageConnection.incoming()).thenReturn(Observable.<Object>just(HELLO_REPLY));
+
+        TransportClient transportClient = mock(TransportClient.class);
+        when(transportClient.connect()).thenReturn(Observable.just(messageConnection));
+
+        ReplicationChannel channel = new SenderReplicationChannel(transportClient, mock(ReplicationChannelMetrics.class));
+
+        return new TestSenderReplicationChannel(channel, id);
+    }
+
+    private static TestSenderReplicationChannel newTimedFailChannel(Integer id, int failAfterMillis) {
+        MessageConnection messageConnection = newMockMessageConnection();
+        when(messageConnection.submit(anyObject())).thenReturn(Observable.<Void>empty());
+        when(messageConnection.incoming()).thenReturn(Observable.<Object>just(HELLO_REPLY));
+
+        TransportClient transportClient = mock(TransportClient.class);
+        when(transportClient.connect()).thenReturn(Observable.just(messageConnection));
+
+        final ReplicationChannel channel = new SenderReplicationChannel(transportClient, mock(ReplicationChannelMetrics.class));
+
+        Observable.empty().delay(failAfterMillis, TimeUnit.MILLISECONDS).subscribe(new Subscriber<Object>() {
```

Comment:
This is intended as a realtime task as it need to sync with some other steps in the test. It'll be difficult to do as a testScheduler unfortunately.